File: leetcode/python/49_group_of_anagrams.py
"""
LeetCode Problem : Group Anagrams
Topic            : Array, Hash Table, String, Sorting
Level            : Medium
URL              : https://leetcode.com/problems/group-anagrams/
"""

from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        """Groups strings that are anagrams of each other from a given list.

        Args:
            strs (List[str]): A list of strings to be grouped.

        Returns:
            List[List[str]]: A list of lists where each inner list contains strings that are anagrams of each other.

        Time Complexity:
            O(N * K): Where N is the length of input array and K is the maximum length of a string in the input.

        Space Complexity:
            O(N * K): Space used to store frequency tuples and grouped anagrams.
        """

        # defaultdict is a subclass of the built-in dict class from the collection module. It is used to provide a default
        # value for a nonexistnt key in the dictionary, eliminating the need for checking if the key exists before using
        # it.
        anagram_map = defaultdict(list)
        for s in strs:
            s_count = [0] * 26
            for char in s:
                s_count[ord(char) - ord("a")] += 1
            key = tuple(s_count)
            anagram_map[key].append(s)
        return list(anagram_map.values())


def run_tests():
    inputs = {"case_1": [["eat", "tea", "tan", "ate", "nat", "bat"]], "case_2": [[""]]}
    outputs = {"case_1": [["eat", "tea", "ate"], ["tan", "nat"], ["bat"]], "case_2": [[""]]}

    solution = Solution()

    for case, input in inputs.items():
        result = solution.groupAnagrams(input[0])
        assert result == outputs[case], f"{case}: expected {outputs[case]}, got {result}"

    print("✅ All tests passed!")


if __name__ == "__main__":
    run_tests()

File: leetcode/go/49_group_of_anagrams.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func groupAnagrams(strs []string) [][]string {
	groups := make(map[string][]string)
	for _, str := range strs {
		signature := generateSignature(str)
		groups[signature] = append(groups[signature], str)
	}

	result := make([][]string, 0, len(groups))
	for _, group := range groups {
		result = append(result, group)
	}

	return result
}

func generateSignature(str string) string {
	alphabets := make([]rune, 26)

	for _, c := range str {
		alphabets[c-'a'] += 1
	}

	var signatureBuilder strings.Builder
	for i, count := range alphabets {
		if count > 0 {
			signatureBuilder.WriteRune('a' + rune(i))
			signatureBuilder.WriteString(fmt.Sprintf("%d", count))
		}
	}

	return signatureBuilder.String()
}

func RunTestGroupAnagrams() {
	// FIX: The test cases still fail because the order of the groups in the result slice result will be randomized
	// as it is converted from a map to a slice.
	testCases := map[string]struct {
		strs   []string
		expect [][]string
	}{
		"case-1": {
			strs: []string{"eat", "tea", "tan", "ate", "nat", "bat"},
			expect: [][]string{
				{"eat", "tea", "ate"},
				{"bat"},
				{"tan", "nat"},
			},
		},
		"case-2": {
			strs: []string{""},
			expect: [][]string{
				{""},
			},
		},
		"case-3": {
			strs: []string{"a"},
			expect: [][]string{
				{"a"},
			},
		},
	}

	for name, testCase := range testCases {
		fmt.Printf("RUN %s\n", name)

		result := groupAnagrams(testCase.strs)
		for i, group := range testCase.expect {
			if !EqualSlices(group, result[i]) {
				fmt.Printf("=== FAILED: expect = %v - got = %v\n", group, result[i])
				os.Exit(1)
			}
		}
		fmt.Printf("=== PASSED\n")
	}

	fmt.Printf("\n✅ All tests passed!\n")
}
